packages/dds/tree/docs/wip/TreeStorage.md

```diff
@@ -0,0 +1,160 @@
+# Tree Storage
+
+This document assumes that the trees will be persisted into some copy on write storage in the form of blobs, which can be fetched using their "handle" which can in turn be serialized into blobs.
+This is the API of Fluid's blobs, but note that fluid requires you to upload the blob to be able to serialize a handle to it.
```

Comment:
For context, this is because Fluid does not know the 'id' of the blob until the upload completes and the resulting id is returned by the service.

There are two alternatives:

1.  We add a level of indirection (e.g., a map from random id -> server blob id) that allows us to serialize the blob in advance of knowing the server's response.

2.  We add a requirement that the service either allow us to specify or precalculate the id.

The former requires maintaining an index indefinitely.  However, by opportunistically replacing the random ids with the final blob ids, we can make the need to load the index rare.

The latter places requirements on the service.

Reply:
Currently the handles seem to be based on hashes of the content, so the option of requiring the service to be able to precalcualte the id seems achievable. I'm not sure if requiring that would be an issue for an servies we have or plan to have. If its not an issue, it would be good for us to leverage it to get ids sooner. If we don't want to put that requirment on the service, we could make services optionally expose it (ex: return the id async, then return the handle, as s two step process, but some services the first step would not ned a network round trip)